src/gsy_e/gsy_e_core/simulation.py

```diff
@@ -68,278 +69,462 @@ def __init__(self):
+        if not_restored_from_state:
+            self.start_time = now(tz=TIME_ZONE)
+            self.paused_time = 0
+
+    def _set_area_current_tick(self, area, current_tick):
+        area.current_tick = current_tick
+        for child in area.children:
+            self._set_area_current_tick(child, current_tick)
+
+    def calculate_total_initial_ticks_slots(self, config, slot_resume, tick_resume, area):
+        """Calculate the initial slot and tick of the simulation, and the total slot count."""
+        slot_count = int(config.sim_duration / config.slot_length)
+
+        if gsy_e.constants.RUN_IN_REALTIME:
+            slot_count = sys.maxsize
+
+            today = datetime.date.today()
+            seconds_since_midnight = time() - mktime(today.timetuple())
+            slot_resume = int(seconds_since_midnight // config.slot_length.seconds) + 1
+            seconds_elapsed_in_slot = seconds_since_midnight % config.slot_length.seconds
+            ticks_elapsed_in_slot = seconds_elapsed_in_slot // config.tick_length.seconds
+            tick_resume = int(ticks_elapsed_in_slot) + 1
+
+            seconds_elapsed_in_tick = seconds_elapsed_in_slot % config.tick_length.seconds
+
+            seconds_until_next_tick = config.tick_length.seconds - seconds_elapsed_in_tick
+
+            ticks_since_midnight = int(seconds_since_midnight // config.tick_length.seconds) + 1
+            self._set_area_current_tick(area, ticks_since_midnight)
+
+            sleep(seconds_until_next_tick)
+
+        if self.slot_length_realtime:
+            self.tick_length_realtime_s = (
+                    self.slot_length_realtime.seconds /
+                    config.ticks_per_slot)
+        return slot_count, slot_resume, tick_resume
+
+    def handle_slowdown_and_realtime(self, tick_no, config):
+        """
+        Handle simulation slowdown and simulation realtime mode, and sleep the simulation
+        accordingly.
+        """
+        if gsy_e.constants.RUN_IN_REALTIME:
+            tick_runtime_s = time() - self.tick_time_counter
+            sleep(abs(config.tick_length.seconds - tick_runtime_s))
+        elif self.slot_length_realtime:
+            current_expected_tick_time = (
+                self.tick_time_counter.add(
+                    seconds=self.tick_length_realtime_s))
+            sleep_time_s = current_expected_tick_time.timestamp() - now().timestamp()
+            if sleep_time_s > 0:
+                sleep(sleep_time_s)
+                log.debug("Tick %s/%s: Sleep time of %s s was applied",
+                          tick_no + 1, config.ticks_per_slot, sleep_time_s)
+
+        self.tick_time_counter = time()
+
+
+@dataclass
+class SimulationSetup:
+    """Static simulation configuration."""
+    seed: int = 0
+    enable_bc: bool = False
+    use_repl: bool = False
+    setup_module_name: str = ""
+    started_from_cli: str = True
+
+    def __post_init__(self):
+        self.seed = self._set_random_seed(self.seed)
```

Comment:
Can you please explain, why this can not be done when initialising the data class?

Reply:
Because the data class constructor only accepts values and cannot execute a function call and then pass the result to the member variable. In order to do the latter, you have to pass the variable you intend to use, and then use the post_init method to call the function that will mutate its value. At least that's what I got from the documentation of the dataclasses.
The obvious fix for this is to ditch the dataclass and use a normal class for that, and admittedly I was also wondering which one of the 2 is the correct solution. Therefore if you would prefer a normal class instead lead me know and I will change it.

Reply:
Thanks for the explanation! I am not sure. I think it is not worth to refactor again, leave like it is. Thanks!